appmgr/src/update/mod.rs

```diff
@@ -0,0 +1,289 @@
+        download_file(&new_label).await?;
+
+        swap_boot_label(&new_label, &mounted_boot).await?;
+        Ok::<_, Error>(())
+    }
+    .await;
+
+    mounted_blue
+        .unmount_label()
+        .await
+        .with_kind(ErrorKind::Filesystem)?;
+    mounted_green
+        .unmount_label()
+        .await
+        .with_kind(ErrorKind::Filesystem)?;
+    mounted_boot
+        .unmount_label()
+        .await
+        .with_kind(ErrorKind::Filesystem)?;
+    potential_error_actions?;
+    Ok(Some(()))
+}
+
+async fn query_mounted_label<'a>(
+    mounted_resource_left: &'a MountedResource<WritableDrives>,
+    mounted_resource_right: &'a MountedResource<WritableDrives>,
+) -> Result<(NewLabel<'a>, CurrentLabel<'a>)> {
+    let output = String::from_utf8(
+        tokio::process::Command::new("cat")
+            .arg("/etc/fstab")
+            .output()
+            .await?
+            .stdout,
+    )?;
+    match &PARSE_COLOR
+        .captures(&output)
+        .ok_or_else(|| anyhow!("Can't find pattern in {}", output))?[1]
+    {
+        x if x == &mounted_resource_left.0.label() => Ok((
+            NewLabel(mounted_resource_left),
+            CurrentLabel(mounted_resource_right),
+        )),
+        x if x == &mounted_resource_right.0.label() => Ok((
+            NewLabel(mounted_resource_right),
+            CurrentLabel(mounted_resource_left),
+        )),
+        e => bail!("Could not find a mounted resource for {}", e),
+    }
+}
+
+async fn download_file(new_label: &NewLabel<'_>) -> Result<(), Error> {
+    let download_request = reqwest::get(URL).await.with_kind(ErrorKind::Network)?;
+    let hash_from_header: String = download_request
+        .headers()
+        .get(HEADER_KEY)
+        .ok_or_else(|| Error::new(anyhow!("No {} in headers", HEADER_KEY), ErrorKind::Network))?
+        .to_str()
+        .with_kind(ErrorKind::InvalidRequest)?
+        .to_owned();
+    let stream_download = download_request.bytes_stream();
+    let file_sum = write_stream_to_label(stream_download, new_label).await?;
+    check_download(&hash_from_header, file_sum).await?;
+    Ok(())
+}
+
+async fn write_stream_to_label(
+    stream_download: impl Stream<Item = Result<rpc_toolkit::hyper::body::Bytes, reqwest::Error>>,
+    file: &NewLabel<'_>,
+) -> Result<Vec<u8>, Error> {
+    let folder = file.0 .0.mount_folder();
+    let file_path = format!("{}/download.img", folder);
+    tokio::process::Command::new("rm")
+        .arg("-rf")
+        .arg(format!("{}/*", folder))
+        .output()
+        .await?;
+    let mut file = tokio::fs::File::create(&file_path)
+        .await
+        .with_kind(ErrorKind::Filesystem)?;
+    let mut hasher = Sha256::new();
+    pin!(stream_download);
+    while let Some(Ok(item)) = stream_download.next().await {
+        file.write(&item).await.with_kind(ErrorKind::Filesystem)?;
+        hasher.update(item);
+    }
+    file.flush().await.with_kind(ErrorKind::Filesystem)?;
```

Comment:
I think I have to flush it, there might be chunks waiting. Need the file written for next command right?